### DISCUSSION TITLE
Mnemonic seed words count

### DISCUSSION THREAD

**Discussion by khaninejad** [pactus-project/pactus#905] [Q&A]
Hi,

I am implementing Mnemonic using the bip39-swift package, but when I input a private key for generating deterministic keys I have got 22 words but the Zarb-go library at validator setup and key generate gives me 12 words! How is that?
https://github.com/bitmark-inc/bip39-swift

Thank you

**Comment by d4rk5h0t** [comment]
@khaninejad 

it looks like the library you are using is faulty the words count according to bip39 standard is [dividable by 3](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#generating-the-mnemonic) there for 22 is out of picture

could you test other library like suggested in official bip39 [doc](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#other-implementations)

**Comment by b00f** [comment]
@d4rk5h0t

Number of seed words are depends of  entropy bytes length. Right now the entropy length is [fixed to 128 bits length](https://github.com/zarbchain/zarb-go/blob/4cf890faec35d56abdef4c6f8849ec3c92edcaf5/cmd/zarb/key/generate.go#L26) which is [enough secure](https://en.bitcoin.it/wiki/Seed_phrase#Explanation)

What we can do is adding new argument to generate command. Default should be 128. User can change it  within the inclusive range of {128, 256} but it should be multiple of 32: 128, 160, 192, 244, 256

**Comment by d4rk5h0t** [comment]
> 
> 
> @d4rk5h0t
> 
> Number of seed words are depends of entropy bytes length. Right now the entropy length is [fixed to 128 bits length](https://github.com/zarbchain/zarb-go/blob/4cf890faec35d56abdef4c6f8849ec3c92edcaf5/cmd/zarb/key/generate.go#L26) which is [enough secure](https://en.bitcoin.it/wiki/Seed_phrase#Explanation)
> 
> What we can do is adding new argument to generate command. Default should be 128. User can change it within the inclusive range of {128, 256} but it should be multiple of 32: 128, 160, 192, 244, 256

@b00f 
you absolutely right i was just making the point
[![image](https://user-images.githubusercontent.com/82785008/118003730-1248f100-b35e-11eb-8d71-dab69e32f586.png)](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#generating-the-mnemonic)

**Comment by khaninejad** [comment]
based on my research library you have used https://github.com/tyler-smith/go-bip39 and test file:
https://github.com/tyler-smith/go-bip39/blob/3423700f970716fa8e0fb7598201e38bfae9ec55/example_test.go
the seeds for "066dca1a2bb7e8a1db2832148ce9933eea0f3ac9548d793112d9a95c9407efad" should be:
"all hour make first leader extend hole alien behind guard gospel lava path output census museum junior mass reopen famous sing advance salt reform"
But I couldn't find a test case to cover this on the zarb-go, my library generates the same as Tylor but I believe you're trimming the seeds to 12 words which are incorrect and should be avoided because of security.
https://iancoleman.io/bip39/#entropy-notes

**Comment by d4rk5h0t** [comment]
>I believe you're trimming the seeds to 12 words which are incorrect and should be avoided because of security.

@khaninejad
there is no trimming happening it's tunable on the library as I describe on image and comment above it depends on your need and accessbility sure enough memorizing 12 words is easier than 21 or so.

actually bip39 is just a protocol to make encryption worthy seed based on re-callable words and as I know best wallets have their own wordlist so there is no guarantee that the words generated will result in same private key seed all over the world so I suggest you have your preference as long as the keys are retrievable on your own end everything is fine

about 12 word it's bare minimum security but sure enough we need tighten it up
